### DISCUSSION TITLE
Contrast Stretch

### DISCUSSION THREAD

**Discussion by CoenraadS** [SixLabors/ImageSharp#1618] [Ideas]
https://imagemagick.org/script/command-line-options.php#contrast-stretch

This is a very useful function that basically takes a histogram, and removes the head/tail % specified by the user and then stretches the remaining range.

Similarly, auto level is basically the same thing just without the clipping

https://imagemagick.org/script/command-line-options.php#auto-level

**Comment by JimBobSquarePants** [comment]
Someone suggested AutoLevel a few years back. See https://github.com/SixLabors/ImageSharp/issues/87

If you'd like to submit an implement of either feature we'd be happy to help you get them completed and merged but there's very little chance anyone on the core team will have the opportunity implement them.

**Reply by CoenraadS** [reply]
Sure, would it be another histogram processor?

e.g. AutoLevelHistogramEqualizationProcessor & ContrastStretchHistogramEqualizationProcessor

**Reply by JimBobSquarePants** [reply]
Yeah, I believe so. I'm not fully clued up on the algorithmic process though. @dlemstra @brianpopow Can you please confirm?

**Reply by brianpopow** [reply]
From reading the description of contrast-stretch and auto-level, I think they are very similar to HistogramEqualization. So I think the naming suggested by @CoenraadS makes sense.

**Reply by JimBobSquarePants** [reply]
Lovely, Thanks @brianpopow

**Reply by CoenraadS** [reply]
Is there a way to automatically get the luminescence depth from a TPixel?

```
var vector = pixelRow[x].ToVector4();
int levels = ???
int luminance = ColorNumerics.GetBT709Luminance(ref vector, levels);
```

**Reply by brianpopow** [reply]
@coenm the luminance levels will be a parameter of the processor, similar as the other HistogramEqualizationProcessors: https://github.com/SixLabors/ImageSharp/blob/master/src/ImageSharp/Processing/Processors/Normalization/HistogramEqualizationProcessor.cs

**Reply by CoenraadS** [reply]
I had been following that file as an example, it's just not a parameter I would expect the user should pass for processors that will work on the full range.

I just found `source.PixelType.BitsPerPixel`, I think that's what I was looking for.
